### ISSUE TITLE
Add-ons cannot be installed from Disco Pane by clicking on the switch button

### ISSUE THREAD

**Issue by ValentinaPC** [mozilla/addons-frontend#2074]
Prerequisites:
- `extensions.webservice.discoverURL` pref is set to point stage/dev

Steps to reproduce:
1. Load about:addons page
2. Try to install an add-on by clicking on the switch button

Expected results:
Add-on is installed.

Actual results:
Add-on cannot be installed.

Notes/Issues:
- add-on can be installed from AMO public pages (i.e https://addons.allizom.org/en-US/firefox/addon/video-downloadhelper/)
- add-ons can be installed from search results pages in about:addons

Verified on FF55(Win 7). Issue is reproducing in AMO-stage and dev.
Video for this issue:
![installdiscopane](https://cloud.githubusercontent.com/assets/15685960/23954086/0b0122a4-099f-11e7-834c-158fe50b41e7.gif)

**Comment by muffinresearch** [comment]
This is caused by the change to disable buttons if the addon is seen to be incompatible. According to the mobile page result for that add-on (on stage [1]) it's incompatible with 54.

[1] https://addons.allizom.org/en-US/firefox/addon/video-downloadhelper/

![add-ons_for_firefox](https://cloud.githubusercontent.com/assets/1514/23960318/8c76645c-099f-11e7-8d90-e6b913b79bce.png)

**Comment by muffinresearch** [comment]
This could be a release blocker if in prod this is going to cause buttons to be disabled in the disco pane when we release.

**Comment by tofumatt** [comment]
Taking a look now.

**Comment by tofumatt** [comment]
For 54 everything seems fine, but Nightly is incompatible with https://addons.mozilla.org/en-GB/firefox/addon/youtube-high-definition/

<img width="110" alt="screenshot 2017-03-15 20 23 13" src="https://cloud.githubusercontent.com/assets/90871/23969177/681b4c94-09bd-11e7-9bb5-4e5aa9794a19.png">

So we should add a message for incompatible add-ons in the disco pane... but we should probably change the add-ons in the disco pane. I think we missed an update but it would be worth cherry-picking into the tag or Nightly will get a greyed-out add-on.

**Comment by mstriemer** [comment]
I thought we only took the minimum version into account in the past? It shows as installable on the desktop site in Nightly.

**Comment by mstriemer** [comment]
(Otherwise every add-on would need a new release for each new Firefox version)

**Comment by tofumatt** [comment]
On the non-mobile site I can't install that add-on with Nightly (v55)–it isn't greyed out but says the add-on is corrupt. So it seems proper that we disable installation on mobile/disco, because it doesn't work. 🤷‍♀️

**Comment by mstriemer** [comment]
It works for me on the production site and I'm on 55 too. Do other add-ons install? Could be you've got some prefs flipped for signing that are messing it up.

I don't think we should prevent an add-on from being installed because it's max version is less than the user's version.

**Comment by tofumatt** [comment]
Ah, shoot, that’s possible, because I’m using Nightly and it’s usually my browser that has weird add-on prefs on. :-/

An add-on with an explicit max version seems like something we should prevent installation of though… otherwise wouldn’t it be compatibility: {max: ‘*’}?

**Comment by ValentinaPC** [comment]
> On the non-mobile site I can't install that add-on with Nightly (v55)–it isn't greyed out but says the add-on is corrupt. So it seems proper that we disable installation on mobile/disco, because it doesn't work. 🤷‍♀️

It works for me using non-mobile site FF55(Win 7); please see video:
![non-mobile](https://cloud.githubusercontent.com/assets/15685960/23984902/bb4fc40a-0a24-11e7-97a8-a2aed5e5c6f1.gif)

**Comment by tofumatt** [comment]
Yeah I must have some weird settings on.

Okay, so we should not block installs for browsers over the max version then?

**Comment by muffinresearch** [comment]
@tofumatt looks like maxVersion should only be taken into account if strictCompatibility is true for old-style add-ons: For old add-ons: https://developer.mozilla.org/en-US/Add-ons/Install_Manifests#targetApplication

For web-ext there's strict_max_version which defaults to '*': https://developer.mozilla.org/en-US/Add-ons/WebExtensions/manifest.json/applications

Looking at the API docs an indication of strict compatibility doesn't appear to be exposed afaict, so if that's the case we have no real way of knowing how to handle versioning in the app.

If that's the case we should probably back out version handling until the API can be updated to indicate if the maxVersion should be relevant or not (assuming that strict versioning isn't happening behind the scenes).

**Comment by tofumatt** [comment]
Ah, shoot. Okay, makes sense. I'll look at backing it out when I'm home at lunchtime. Feel free to do it earlier if that's too late ;-)

Matthew Riley MacPherson (sent from mobile)

**Comment by diox** [comment]
There are a few things going on, here is a summary of my findings:

**TL;DR**: if you want to follow what AMO desktop page do, ignore max version, implement application checking and min version checking.

- addons-server legacy disco pane and legacy API (that Firefox's AddonManager uses) obey a special compat mode parameter passed through the URL. Firefox chooses what parameter to send depending on the values of the `extensions.strictCompatibility` and `extensions.checkCompatibility` preferences that you can set in `about:config`. That parameter can be "ignore", "strict" or "normal" (which is the default).
  * In strict mode, it looks at both the min and max version on the last public version of the addon
  * On ignore mode, it does the same thing but only looks at the min version
  * In normal mode, it tries harder to find a compatible version, by looking at previously public versions as well, only obeying max version if strict compatibility is set on the file corresponding to that version, or if the file has binary components. It does some very nasty queries to achieve that.

  The rest of the AMO website does not implement that behaviour at all.

- addons-server detail page only cares about what applications the add-on support and for which min version. The max version and strict compatibility are ignored completely. (That's a simplification: there is code to handle super old versions like Firefox 4.0, but I'm ignoring this here).

I think we should follow the detail page behaviour here. *Maybe* also implementing strict compatibility checking for max version. The different modes the legacy API and disco support are quite complex and lead to weird situations where users can be stuck on old versions of an add-on, I don't think we should implement that.

In addition:
- addons-server search page has a `appver` query parameter but does weird things with it. It will use it to filter out add-ons from results using the min/max compatibility info on their current public version, but *only* when searching for non-extensions (e.g. complete themes, search tools, language packs etc) explicitly.

- addons-server API has a `appversion` query parameter and uses it to filter out add-ons from results using the min/max compatibility info on their current public version. I am not sure whether or not I should fix it to do what the current search pages do or not.

**Comment by vcarciu** [comment]
Verified as fixed : 
![radiobuttons](https://cloud.githubusercontent.com/assets/1583842/24008924/f092476c-0a7b-11e7-9b50-a16e58170f44.gif)

**Comment by vcarciu** [comment]
This was now pushed to Stage and it is working here too
